Minimal encoding for Data Matrix with multi ECI and GS1-FNC1 support

AlexGeller1 opened the pull request:
Feature is off by default and can be activated by EncodeTypeHint.DATA_MATRIX_COMPACT

srowen commented on core/src/main/java/com/google/zxing/datamatrix/decoder/DecodedBitStreamParser.java:
```diff
@@ -555,4 +588,70 @@ private static int unrandomize255State(int randomizedBase256Codeword,
     return tempVariable >= 0 ? tempVariable : tempVariable + 256;
   }
 
+  private static final class ECIStringBuilder {
```
Let's refactor this into a package-private class in the .common package and reuse it across this and QR codes

AlexGeller1 replied:
Just to be sure, in this PR I did not modify decoder code like this class. The commits appear in this PR but they are really two weeks old (I am still struggling with GitHub and occasionally this happens that it is shown like this).
It you are OK with it, then I can look in a separate PR if the ECIStringBuilder class has common code with the QR-Code decoder. I did however now move some code that is common to the two encoders into the common package.

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+  }
+
+  static boolean isExtendedASCII(char ch, int fnc1) {
+
```
Fix up the code style warnings noted in the automated checks and things like removing a blank line here

AlexGeller1 replied:
I fixed that line but I failed to find the spot in GitHub where the automated check findings are displayed. There are 3 checks for Java down below (CodeQL Analyze, LGTM and Code scanning results / CodeQL). Is it one of them? I clicked around in the "details" of these but couldn't find any warnings. Can you point me to it?

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+  }
+
+  private static boolean isInC40Shift2Set(char ch, int fnc1) {
+    final char[] chars = {'!', '"', '#', '$', '%', '&', '\'', '(', ')', '*',  '+', ',', '-', '.',
```
private static final member?

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+  public static String encodeHighLevel(String msg, Charset priorityCharset, int fnc1, SymbolShapeHint shape) {
+    if (msg.startsWith(HighLevelEncoder.MACRO_05_HEADER) && msg.endsWith(HighLevelEncoder.MACRO_TRAILER)) {
+      //TODO: Add support for Macro 5 (236)
+      //@Sean: Is a similar syntax used to say that is it GS1 input? Perhaps we should use it?
```
Space after //
I'm not sure about the answer here

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+    characterLength[0] = 0;
+    return 0;
+  }
+  static void addEdges(Input input, Edge[][] edges, int from, Edge previous) {
```
Blank line

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+      }
+  
+      if (input.haveNCharacters(from,3) && HighLevelEncoder.isNativeX12(input.charAt(from)) &&
+                                                 HighLevelEncoder.isNativeX12(input.charAt(from + 1)) &&
```
Indent is too far - break down to a line per clause

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+      addEdge(edges, new Edge(input, Mode.B256, from, 1, previous));
+    }
+
+//We create 4 EDF edges,  with 1, 2 3 or 4 characters length. The fourth normally doesn't have a latch to ASCII unless
```
Wrong indent

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+  }
+
+  private static final class Edge {
+    static final int[] allCodewordCapacities = {3, 5, 8, 10, 12, 16, 18, 22, 30, 32, 36, 44, 49, 62, 86, 114, 144, 174,
```
private?

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+ * X12 -> ASCII: word(c1,c2,c3), 254
+ * EDIFACT -> ASCII: Unlatch character,0,0,0 or c1,Unlatch character,0,0 or c1,c2,Unlatch character,0 or 
+ * c1,c2,c3,Unlatch character
+ * B256 -> ASCII: without latch after n bytes
```
Wrong indent

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+      this.input = solution.input;
+      int length = 0;
+      int size = 0;
+      ArrayList<Byte> bytesAL = new ArrayList();
```
List for ref type, not ArrayList

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+        bytes[i] = bytesAL.get(i).byteValue();
+      }
+    }
+    static int prepend(byte[] bytes, ArrayList<Byte> into) {
```
Blank lines between methods

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+      return bytes.length;
+    }
+  }
+  private static final class Input {
```
Blank line

srowen commented on core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java:
```diff
@@ -0,0 +1,1395 @@
+      return bytes;
+    }
+    /**
+     * returns the size in bytes
```
`@return`

srowen reviewed (changes requested):
Can any of the new encoder code be shared with QR code, in a class in .common?

srowen commented:
Regarding your last comment, what do you mean about github issues? your previous commits should be merged. do you need to rebase your branch? in any event the idea is generally to share whatever code can be shared rather than copy it

srowen commented:
Click "View on LGTM" - https://lgtm.com/projects/g/zxing/zxing/rev/pr-df909951cf4b96cc4280c40be1f9d86827069744

AlexGeller1 commented:
> Regarding your last comment, what do you mean about github issues? 

This PR lists for example this: "Added support for multiple character-ECIs" but this commit is two weeks old and already merged into HEAD.

>your previous commits should be merged. do you need to rebase your branch?

There is some problem there. I apparently don't rebase correctly and then it shows things this way.
 
>in any event the idea is generally to share whatever code can be shared rather than copy it

This remark I don't understand.

AlexGeller1 commented:
> Click "View on LGTM" - https://lgtm.com/projects/g/zxing/zxing/rev/pr-df909951cf4b96cc4280c40be1f9d86827069744

Thanks, I thought that I had addressed the issues listed in this one or are you pointing me to it because it is also the place where the warning about the missing space is found?

srowen commented:
OK, just rebase this branch on master and force-push this branch and see if it makes more sense.
To be concrete: can we have one ECIStringBuilder class, not two?

AlexGeller1 commented:
> OK, just rebase this branch on master and force-push this branch and see if it makes more sense. To be concrete: can we have one ECIStringBuilder class, not two?

There is only one ECIStringBuilder class. It is an inner class of DecodedBitStreamParser. In this pull request I don't want to make any changes to DecodedBitStreamParser. I now rebased on master and committed the master version of the file so that when this PR gets merged into master, there will be no changes for this file.

srowen commented:
There is an ECIStringBuilder in the master branch already, but this still appears to add one. Are you sure it's rebased and pushed? https://github.com/zxing/zxing/blob/c4266bf09c1a3d13c236a8e3a15f8e34cbda025a/core/src/main/java/com/google/zxing/datamatrix/decoder/DecodedBitStreamParser.java#L591

AlexGeller1 commented:
> There is an ECIStringBuilder in the master branch already, but this still appears to add one. Are you sure it's rebased and pushed?

Yes, I verified just now once more to be sure. The file core/src/main/java/com/google/zxing/datamatrix/decoder/DecodedBitStreamParser.java in this PR is identical with the master branch version.

srowen commented:
This code already exists in master - your branch does not have commits from master, from about Dec 25 onwards. You need to pull the master branch, rebase your branch on it, and force-push the branch

AlexGeller1 commented:
> This code already exists in master - your branch does not have commits from master, from about Dec 25 onwards. You need to pull the master branch, rebase your branch on it, and force-push the branch

Puh, I hear your words but I can't figure out how it maps to git commands. Here is an example of what I have been trying:
alex@AlexBuster:~/ZXing/zxing$ git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
alex@AlexBuster:~/ZXing/zxing$ git checkout feat-data-matrix-eci-and-compact-encoding
Switched to branch 'feat-data-matrix-eci-and-compact-encoding'
Your branch is up to date with 'origin/feat-data-matrix-eci-and-compact-encoding'.
alex@AlexBuster:~/ZXing/zxing$ 
alex@AlexBuster:~/ZXing/zxing$ git rebase origin/master -X theirs
First, rewinding head to replay your work on top of it...
Auto-merging core/src/main/java/com/google/zxing/datamatrix/decoder/DecodedBitStreamParser.java
[detached HEAD c175f1cc1] Added support for multiple character-ECIs
 Date: Sat Dec 25 17:40:58 2021 +0100
 1 file changed, 15 insertions(+), 25 deletions(-)
Committed: 0001 Added support for multiple character-ECIs
[detached HEAD b115c2143] - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
 Date: Sun Dec 26 18:43:39 2021 +0100
 1 file changed, 16 insertions(+), 8 deletions(-)
Committed: 0002 - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
[detached HEAD ed707a46b] Improved performance for the general case (without ECI)
 Date: Mon Dec 27 16:06:45 2021 +0100
 1 file changed, 17 insertions(+), 15 deletions(-)
Committed: 0003 Improved performance for the general case (without ECI)
[detached HEAD c64700b6f] - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
 Date: Sun Dec 26 18:43:39 2021 +0100
 1 file changed, 1 insertion(+)
Committed: 0004 - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
[detached HEAD b834fc50d] Improved performance for the general case (without ECI)
 Date: Mon Dec 27 16:06:45 2021 +0100
 1 file changed, 1 deletion(-)
Committed: 0005 Improved performance for the general case (without ECI)
[detached HEAD 5d20c5789] Added Data Matrix minimal encoding with multi-ECI and GS1-FNC1 support.
 Date: Mon Jan 10 10:16:20 2022 +0100
 5 files changed, 1563 insertions(+), 11 deletions(-)
 create mode 100755 core/src/main/java/com/google/zxing/datamatrix/encoder/MinimalEncoder.java
Committed: 0006 Added Data Matrix minimal encoding with multi-ECI and GS1-FNC1 support.
[detached HEAD 63a18e978] - Removed useless null checks - Fixed javadoc param
 Date: Mon Jan 10 13:30:55 2022 +0100
 2 files changed, 7 insertions(+), 7 deletions(-)
Committed: 0007 - Removed useless null checks - Fixed javadoc param
[detached HEAD 056bf0698] - Moved common encoder code into common.ECIEncoderSet. - Added multi ECI test case - Performed various code style changes
 Date: Mon Jan 10 17:27:45 2022 +0100
 4 files changed, 301 insertions(+), 307 deletions(-)
 create mode 100755 core/src/main/java/com/google/zxing/common/ECIEncoderSet.java
Committed: 0008 - Moved common encoder code into common.ECIEncoderSet. - Added multi ECI test case - Performed various code style changes
[detached HEAD a8efe6482] - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
 Date: Sun Dec 26 18:43:39 2021 +0100
 1 file changed, 1 insertion(+)
Committed: 0009 - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
[detached HEAD e4aaeba9b] Improved performance for the general case (without ECI)
 Date: Mon Dec 27 16:06:45 2021 +0100
 1 file changed, 1 deletion(-)
Committed: 0010 Improved performance for the general case (without ECI)
[detached HEAD 060ba9423] - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
 Date: Sun Dec 26 18:43:39 2021 +0100
 1 file changed, 1 insertion(+)
Committed: 0011 - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
[detached HEAD 6953bb262] Improved performance for the general case (without ECI)
 Date: Mon Dec 27 16:06:45 2021 +0100
 1 file changed, 1 deletion(-)
Committed: 0012 Improved performance for the general case (without ECI)
[detached HEAD 064acdfb2] - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
 Date: Sun Dec 26 18:43:39 2021 +0100
 1 file changed, 1 insertion(+)
Committed: 0013 - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
[detached HEAD 2e25ccdcb] Improved performance for the general case (without ECI)
 Date: Mon Dec 27 16:06:45 2021 +0100
 1 file changed, 1 deletion(-)
Committed: 0014 Improved performance for the general case (without ECI)
[detached HEAD e23146c7f] - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
 Date: Sun Dec 26 18:43:39 2021 +0100
 1 file changed, 1 insertion(+)
Committed: 0015 - Changed from ArrayList<Byte> to ByteArrayOutputStream - Added blank lines
[detached HEAD 886e46d71] Improved performance for the general case (without ECI)
 Date: Mon Dec 27 16:06:45 2021 +0100
 1 file changed, 1 deletion(-)
Committed: 0016 Improved performance for the general case (without ECI)
[detached HEAD 8e6442545] Added support for multiple character-ECIs
 Date: Sat Dec 25 17:40:58 2021 +0100
 1 file changed, 15 insertions(+), 25 deletions(-)
Committed: 0017 Added support for multiple character-ECIs
[detached HEAD 58772d350] Updating DecodedBitStreamParser to origin/head
 Date: Tue Jan 11 00:05:26 2022 +0100
 2 files changed, 46 insertions(+), 36 deletions(-)
Committed: 0018 Updating DecodedBitStreamParser to origin/head
[detached HEAD 573e72252] - Updating DecodedBitStreamParser to origin/master version - Whitespace edits
 Date: Tue Jan 11 00:07:15 2022 +0100
 1 file changed, 5 insertions(+), 4 deletions(-)
Committed: 0019 - Updating DecodedBitStreamParser to origin/master version - Whitespace edits
All done.
alex@AlexBuster:~/ZXing/zxing$ git status
On branch feat-data-matrix-eci-and-compact-encoding
Your branch and 'origin/feat-data-matrix-eci-and-compact-encoding' have diverged,
and have 20 and 19 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

nothing to commit, working tree clean
alex@AlexBuster:~/ZXing/zxing$ git pull
First, rewinding head to replay your work on top of it...
Applying: Add support for multiple character-ECIs in data matrix decoder (#1474)
Using index info to reconstruct a base tree...
M	core/src/main/java/com/google/zxing/datamatrix/decoder/DecodedBitStreamParser.java
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
alex@AlexBuster:~/ZXing/zxing$ git status
On branch feat-data-matrix-eci-and-compact-encoding
Your branch is up to date with 'origin/feat-data-matrix-eci-and-compact-encoding'.

nothing to commit, working tree clean

srowen commented:
Given your setup, after you checkout master, you need to git pull, then git push origin master. You're not getting updates from master to rebase on.

srowen commented:
Sorry I mean you will need to pull from the main repo. If you have github.com/zxing/zxing as a remote call 'upstream', then it's git pull upstream master

AlexGeller1 commented:
> Sorry I mean you will need to pull from the main repo. If you have github.com/zxing/zxing as a remote call 'upstream', then it's git pull upstream master

Thanks for your help. I am learning. It went as follows:
```code
alex@AlexBuster:~/ZXing/zxing$ git pull upstream master
From https://github.com/AlexGeller1/jsweet
 * branch                master     -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
Generating patches: 100% (3549/3549), done.
Applying: 
Applying: 
Applying: 
Applying: 
Applying: 
Applying: Fixed dumb bug that caused grid sampling to sample 2x too wide in the horizontal direction.
Applying: Initial revision
Applying: Added more javadoc
Applying: More javadoc
Applying: More javadoc
Applying: More javadoc
Applying: More javadoc
Applying: More javadoc
Applying: Added more test cases
Applying: Fix compile problem for JDK 1.4
Applying: Oops, bug fix for MaskPattern100
Applying: Adjust tolerance of method that decides whether a finder/alignment pattern has been crossed so that it passes on low resolution images more correctly -- just a matter of resorting to some floating point math.
Applying: Should fix up remaining problems in J2ME build and enable build of deployable .jar file
Applying: Added icon
Applying: Replace JAIPerspectiveTransform with PerspectiveTransform
Applying: Better, still some bugs
Applying: Fixed the rest of my typos and added basic unit test
Applying: Tiny tweak
Applying: Break out properties that user should customize into a build.properties file
Applying: Added the initial version of my UPC reader and modified some common files
Applying: Added top-level build file, removed local copy of JUnit, added test target to core
Applying: "Fire" button now triggers decoding
Applying: Code tweaks and so forth with Daniel
Applying: Added copyright statement

```
... hundreds of similar lines ...
```code

Applying: Fix <br/> escaping that doesn't work
Applying: A few inspection cleanups
Applying: Update and refactor poms under one parent
Applying: import https://mattmccutchen.net/bigint/
Applying: compile .cc as well .cpp files to include bigint
Applying: pdf417 patch sans vs tweaks
Applying: c++ port of r2660
Applying: Issue 1670 remove Ant scripts in favor of Maven where possible
Applying: make VC++ happy with pdf417
Applying: xcode project file udpates and a portability fix
Using index info to reconstruct a base tree...
Falling back to patching base and 3-way merge...
Auto-merging .gitignore
CONFLICT (add/add): Merge conflict in .gitignore
error: Failed to merge in the changes.
Patch failed at 2363 xcode project file udpates and a portability fix
hint: Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

alex@AlexBuster:~/ZXing/zxing$ vi .gitignore
alex@AlexBuster:~/ZXing/zxing$ git add .gitignore
alex@AlexBuster:~/ZXing/zxing$ git rebase --continue
Applying: xcode project file udpates and a portability fix
Applying: reflect the ant->mvn changes in the gitignore config
Using index info to reconstruct a base tree...
M	.gitignore
Falling back to patching base and 3-way merge...
Auto-merging .gitignore
CONFLICT (content): Merge conflict in .gitignore
error: Failed to merge in the changes.
Patch failed at 2364 reflect the ant->mvn changes in the gitignore config
hint: Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

alex@AlexBuster:~/ZXing/zxing$ vi .gitignore
alex@AlexBuster:~/ZXing/zxing$ git add .gitignore
alex@AlexBuster:~/ZXing/zxing$ git rebase --continue
Applying: reflect the ant->mvn changes in the gitignore config
No changes - did you forget to use 'git add'?
f there is nothing left to stage, chances are that something else
already introduced the same changes; you might want to skip this patch.

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

alex@AlexBuster:~/ZXing/zxing$ git rebase --skip
Applying: Copy .jar files on package, not install
Applying: Make C++ consistent with Java for pdf417
Applying: fix wrong abs() call and modernize formatting
Applying: fix various wrong abs; other cleanup
Applying: between case-ind fs and recursive includes, osx is confused
Applying: tweak for a gcc variant

```
... many more conflicts. At some point there where rename conflicts so I finally just skipped them ..
```code

Removing actionscript/core/build.xml
Removing actionscript/core/.settings/org.eclipse.core.resources.prefs
Removing actionscript/core/.project
Removing actionscript/core/.flexLibProperties
Removing actionscript/core/.actionScriptProperties
Auto-merging README.md
CONFLICT (content): Merge conflict in README.md
error: Failed to merge in the changes.
Patch failed at 2832 Retire actionscript
hint: Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

alex@AlexBuster:~/ZXing/zxing$ git status
rebase in progress; onto 2e8800ded
You are currently rebasing branch 'master' on '2e8800ded'.
  (fix conflicts and then run "git rebase --continue")
  (use "git rebase --skip" to skip this patch)
  (use "git rebase --abort" to check out the original branch)

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	deleted:    actionscript/core/.actionScriptProperties
	deleted:    actionscript/core/.flexLibProperties
	deleted:    actionscript/core/.project
	deleted:    actionscript/core/.settings/org.eclipse.core.resources.prefs
	deleted:    actionscript/core/build.xml
	deleted:    actionscript/core/src/com/google/zxing/BarcodeFormat.as
	deleted:    actionscript/core/src/com/google/zxing/Binarizer.as
	deleted:    actionscript/core/src/com/google/zxing/BinaryBitmap.as
	deleted:    actionscript/core/src/com/google/zxing/BufferedImageLuminanceSource.as
```
...
```code

	deleted:    actionscript/zxing client/html-template/history/history.css
	deleted:    actionscript/zxing client/html-template/history/history.js
	deleted:    actionscript/zxing client/html-template/history/historyFrame.html
	deleted:    actionscript/zxing client/html-template/index.template.html
	deleted:    actionscript/zxing client/html-template/playerProductInstall.swf
	deleted:    actionscript/zxing client/src/.actionScriptProperties
	deleted:    actionscript/zxing client/src/.flexProperties
	deleted:    actionscript/zxing client/src/.project
	deleted:    actionscript/zxing client/src/zxing_client.mxml

Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add <file>..." to mark resolution)

	both modified:   README.md

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	.travis.yml~HEAD
	target/
	zxingorg/target/

alex@AlexBuster:~/ZXing/zxing$
```

pzygielo commented:
```
alex@AlexBuster:~/ZXing/zxing$ git pull upstream master
From https://github.com/AlexGeller1/jsweet
                                    ^^^^ ?
```
1. I believe there is convention, not a must, to have `upstream` to point to original repository, not a fork.
2. In case presented above  - upstream points to very different repository, which does not seem to be closely related to zxing.

srowen commented:
Did you pull upstream onto your branch? no you don't want to do that.

The standard setup should be: `origin` remote pointing to your fork of the project, and `upstream` remote pointing to this project. Your `origin/master` should stay in sync with `upstream/master` and not deviate. So you can at any time checkout master, pull from upstream, push to origin. Then in your branch, as needed you can `git rebase master` after pulling master updates. I would recommend rebase over any merges, as I think it's clearer.

AlexGeller1 commented:
> Did you pull upstream onto your branch? no you don't want to do that.
> 
How about closing this PR and I start a fresh one?

srowen commented:
That's fine if it's the easiest path. Just pull upstream master before branching again. There are ways to revert anything you don't want on your branch then rebase but could be easier to start over for relatively simple changes

AlexGeller1 commented:
> ```
> alex@AlexBuster:~/ZXing/zxing$ git pull upstream master
> From https://github.com/AlexGeller1/jsweet
>                                     ^^^^ ?
> ```
> 
> 1. I believe there is convention, not a must, to have `upstream` to point to original repository
> 2. In case presented above  - upstream points to very different repository, which does not seem to be closely related to zxing.

@pzygielo Thanks a ton. I think I have fixed that now.
```code
alex@AlexBuster:~/ZXing/zxing$ git checkout master
Already on 'master'
Your branch is up to date with 'origin/master'.
alex@AlexBuster:~/ZXing/zxing$ git remote set-url upstream https://github.com/AlexGeller1/zxing.git
alex@AlexBuster:~/ZXing/zxing$ git pull upstream master
From https://github.com/AlexGeller1/zxing
 * branch                master     -> FETCH_HEAD
Already up to date.
Current branch master is up to date.
```

srowen commented:
You want origin to be https://github.com/AlexGeller1/zxing.git, and upstream to be https://github.com/zxing/zxing.git

AlexGeller1 commented:
Very good, thanks!
```code
alex@AlexBuster:~/ZXing/zxing$ git remote set-url upstream https://github.com/zxing/zxing.git
alex@AlexBuster:~/ZXing/zxing$ git pull upstream master
remote: Enumerating objects: 10, done.
remote: Counting objects: 100% (10/10), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 10 (delta 4), reused 10 (delta 4), pack-reused 0
Unpacking objects: 100% (10/10), done.
From https://github.com/zxing/zxing
 * branch                master     -> FETCH_HEAD
   c4266bf09..491006ac3  master     -> upstream/master
Updating c4266bf09..491006ac3
Fast-forward
 core/src/main/java/com/google/zxing/NotFoundException.java | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
Current branch master is up to date.

AlexGeller1 commented:
Thank you for your patient explaining. I created the new PR.
